migrate.py

```diff
@@ -200,24 +199,26 @@ def _must_keep(self, snapshots: List[str], snapshot: str, keep_last_n: Optional[
             if keep_monthly_n is None:
                 return True
             monthly_snapshots = [snapshot for snapshot in snapshots if self._is_monthly(snapshots, snapshot)]
-            if self._is_snapshot_in_tail_of_list(monthly_snapshots, snapshot, keep_monthly_n):
+            if self._is_among_n_newest(monthly_snapshots, snapshot, keep_monthly_n):
                 return True
 
         return False
 
     def _find_next_snapshot(self, dataset_name: str, snapshots: List[Dict[str, str]], snapshots_in_restic: Dict[str, str],
                             keep_last_n: Optional[int], keep_weekly_n: Optional[int], keep_monthly_n: Optional[int]) -> Optional[str]:
         """
-        We consider each snapshot in `snapshots` in the given order, so it should be sorted by creation time.
+        `snapshots` must be sorted by creation time.
         """
-        snapshots_with_size = [snapshot for snapshot in snapshots if snapshot["used"] != 0]
         snapshot_names_in_restic = set([s["name"] for s in snapshots_in_restic])
-        for snapshot in snapshots:
+        for i, snapshot in reversed(list(enumerate(snapshots))):
             snapshot_name = snapshot["name"]
-            if snapshot["used"] == 0:
-                print(F"Skipping snapshot {dataset_name}@{snapshot_name} because of zero size.")
-                continue
-            if not self._must_keep(snapshots_with_size, snapshot, keep_last_n, keep_weekly_n, keep_monthly_n):
+            if i > 0 and snapshots[i-1]["used"] == 0:
+                parent_name = snapshots[i-1]["name"]
+                diff_output = _eval(f"zfs diff {dataset_name}@{parent_name} {dataset_name}@{snapshot_name} 2>&1 | head -c1 | wc -c")
+                if diff_output == "0\n":
+                    print(F"Skipping snapshot {dataset_name}@{snapshot_name} because of zero diff.")
+                    continue
+            if not self._must_keep(snapshots_in_restic, snapshot, keep_last_n, keep_weekly_n, keep_monthly_n):
```

Comment:
I don't really understand why we can use `snapshots_in_restic` here.

Reply:
Because the script now iterates backwards through the ZFS snapshots. This line does assume that the restic repo is not shared with other datasets, though.